Module quality update

Joris29 opened the pull request:
#### Pull Request (PR) description
This PR will make several changes to follow the best practices of puppet more and remove deprecated parameters.

Like removing params.pp
Removing projects_storage_type parameter
...

This will be a breaking change for some use cases.

#### This Pull Request (PR) fixes the following issues
Fixes #197
Fixes #373
Makes #398 obsolete
Fixes #406
Fixes #426
Fixes #445
Makes #447 obsolete
Fixes #451
Fixes #452
Fixes #457
Fixes #472
Fixes #479
Fixes #484
Makes #487 obsolete
Makes #521 obsolete

Joris29 commented:
@kenyon Hello it's been a while, in my attempt to remove the params.pp i also took the opportunity to rework the module entirely.

The unit tests still need to updated and maybe some other things not related to the code.
The code has been tested on a RHEL 8 machine.

My main questions is how to progress towards a new release is this something i need to do or how would this be done?

kenyon reviewed (commented):
Looks pretty good.

We like to have parameter defaults documented in REFERENCE.md as much as possible, so move everything from `common.yaml` to `init.pp`.

Don't put anything in `spec/fixtures`, the test code automatically copies everything there.

I would avoid aligning the `=` signs in class parameter lists, since it causes lots of diffs which makes changes harder to review, and can get hard to read when lines get lengthy.

As far as a new release, it's the usual procedure: the change is merged into the `master` branch, and eventually we make a new major release.

Joris29 commented:
Good to know nothing has to be in spec/fixtures

I will try to move some things from common but what about big hashes like the admin policy and the os family specific things?

kenyon commented:
OS-specific things should still go in the OS data file in hiera. I think big hashes in `init.pp` are fine, this allows them to be documented in `REFERENCE.md`.

Joris29 commented:
I moved the parameter values to init.pp but i'm having some difficulties with framework_config.

This should be merged with the defaults from rundeck because if you overwrite the framework_config paramter then you lose all defaults if you don't set them explicitly like rdeck.base and i would like to avoid this so what's the best approach for this?

kenyon commented:
> I moved the parameter values to init.pp but i'm having some difficulties with framework_config.
> 
> This should be merged with the defaults from rundeck because if you overwrite the framework_config paramter then you lose all defaults if you don't set them explicitly like rdeck.base and i would like to avoid this so what's the best approach for this?

One thing you can do is make `$framework_config` only override the defaults. Set the defaults in the code, then merge (`+`) the hashes.

Joris29 commented:
@kenyon I updated the specs and references even further for me everything checks out one check fails but it has nothing to do with the actual unit tests. The only thing still on my todo is to update the readme as this includes wrong example for now.
Feel free to add more todo's if needed i also updated which issues/PR's will be fixed or become obsolete i might have forgot some but this can resolved later i guess?

kenyon commented on manifests/init.pp:
```diff
@@ -1,244 +1,241 @@
-  Hash[String,String]                 $storage_encrypt_config             = {},
-) inherits rundeck::params {
-  validate_rd_policy($acl_policies)
+  Stdlib::Absolutepath           $override_dir,
```
I would also eliminate the alignment spaces between the data type and variable name, just like you did between the variable and the default value.

We had a discussion about this at https://github.com/voxpupuli/voxpupuli.github.io/pull/322.

kenyon commented on manifests/init.pp:
```diff
@@ -1,244 +1,241 @@
+  Stdlib::Absolutepath           $override_dir,
+  Hash                           $repo_config,
+  Boolean                        $manage_repo = true,
+  String                         $package_ensure = 'installed',
```
We like to enforce nonempty strings wherever it makes sense. Example:

```suggestion
  String[1] $package_ensure = 'installed',
```

kenyon commented on data/Debian.yaml:
```diff
@@ -0,0 +1,14 @@
+---
+rundeck::override_dir: '/etc/default'
+
+rundeck::repo_config:
+  'apt::source':
+    'rundeck':
+      location: 'https://packagecloud.io/pagerduty/rundeck/any'
+      release: 'any'
+      repos: 'main'
+      comment: 'Official repository for Rundeck'
+      key:
+        id: '0DDD2FA79B15D736ECEA32B89B5206167C5C34C0'
+        source: 'https://packagecloud.io/pagerduty/rundeck/gpgkey'
+        server: 'keyserver.ubuntu.com'
```
Use modern APT keyring, the keyserver way is deprecated.

Also remove the `apt::source` level so that you can just iterate over `rundeck::repo_config`. Same for `RedHat.yaml`.

```suggestion
rundeck::repo_config:
  rundeck:
    location: https://packagecloud.io/pagerduty/rundeck/any
    release: any
    repos: main
    comment: Official repository for Rundeck
    key:
      name: rundeck.asc
      source: https://packagecloud.io/pagerduty/rundeck/gpgkey
```

kenyon commented on manifests/install.pp:
```diff
@@ -1,101 +1,54 @@
-          before        => Package['rundeck'],
+    /RedHat|Debian/: {
+      if $rundeck::manage_repo {
+        $rundeck::repo_config.each() | String $_resource_type, Hash $_resources | {
+          if downcase($_resource_type) == 'apt::source' {
+            Class['Apt::Update'] -> Package['rundeck']
+          }
+          create_resources($_resource_type, $_resources, { 'before' => Package['rundeck'] })
```
Simplify, and eliminate `create_resources`.

```suggestion
        $rundeck::repo_config.each |String $_apt_source, Hash $_attrs| {
          apt::source { $_apt_source:
            * => $_attrs,
            before => Package['rundeck'],
          }
```

jay7x replied:
if @kenyon's suggestion about removing the `apt::source` and `yumrepo` from the hiera hash will be implemented then this part should be split for RH/Debian families again.

Joris29 replied:
Yep working on it will split them.

Joris29 replied:
Updated the loop not optimal in my opinion because you have some resource twice with the same value but i will try to optimise this tomorrow.
I wonder why the debian tests fail with: Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, key hash must contain at least an id entry (file: /etc/puppetlabs/code/environments/production/modules/apt/manifests/source.pp, line: 126, column: 9)

The syntax seems correct and valid according to the apt module

kenyon replied:
Ah that's because the acceptance tests are installing puppetlabs-apt v9.1.0, but we need v9.2.0 for the modern keyrings feature: https://github.com/puppetlabs/puppetlabs-apt/pull/1150

Joris29 replied:
Ah ok that explains should i wait then and update acceptance tests with new apt release?

kenyon replied:
The acceptance tests should use the new version automatically once it's released.

Joris29 replied:
Alright then i will wait for the apt PR and try to simplify the os family case structure and update the readme a bit more

kenyon commented on manifests/service.pp:
```diff
@@ -1,33 +1,28 @@
       ensure  => file,
       mode    => '0644',
-      content => template($service_config),
+      content => template($rundeck::service_config),
```
Make these all EPP instead of ERB?

```suggestion
      content => epp($rundeck::service_config),
```

kenyon reviewed (commented):
Looks pretty good so far.

kenyon reviewed (approved):
Looks good.

Do you want to squash the commits?

Would it be helpful to add a section to the README about how to migrate to this new version of the module from the current version?

Joris29 commented:
Yeah maybe it's best to squash the commits as there are quite a lot and some commits are not really clear on what they do exactly.

For the migration part i think this will kind off be part of the change logs because all the breaking changes will be in there.
There is also not much to it except renaming/removing parameters or restructuring some data like the auth_config or key_storage_config but examples for both are added to the readme.

kenyon commented:
Release PR #523 includes this change.

Fabian1976 commented:
I'm sorry to say that this pull request actually reduced the quality of the module.

This parameter:
```
dataSource.dbCreate = update
```
Found on [this](https://github.com/voxpupuli/puppet-rundeck/blob/33879eb96e6ea2125a22c2670ea38c15bc19e017/templates/rundeck-config.properties.epp#L14C1-L14C29) line, is now hard-coded in the module and prevents new installation of rundeck and a postgresql backend.

In the version before this change (v8.0.1), it was a proper parameter:
```
dataSource {
  dbCreate = "<%= @database_config['dbCreate'] %>"
```

Because of that fixed parameter, new installs of rundeck with Postgresql generate this error:
```
[2022-09-29T21:22:05,759] ERROR boot.SpringApplication - Application run failed
liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for change set core/ExecReportJcExecIdToExecutionId.groovy::1653344096108-52::gschueler (generated):
     Reason: liquibase.exception.DatabaseException: ERROR: column "jc_exec_id" does not exist
  Position: 44 [Failed SQL: (0) update base_report set execution_id = cast(JC_EXEC_ID as bigint)]
```

Manually removing this parameter results in rundeck starting just fine.

I will also log an issue for this.

hseffler commented:
This PR has removed a large number of options from the rundeck class.

E.g. I now get:

```
  has no parameter named 'auth_types'
  has no parameter named 'acl_policies'
  has no parameter named 'session_timeout'
  parameter 'database_config' unrecognized key 'type'
  parameter 'database_config' unrecognized key 'dbCreate'
```

Is there any documentation on how to migrate without digging through the whole module?
